plugins/cryptolib/cryptolib.cpp

```diff
@@ -0,0 +1,669 @@
+void doPKISign(size32_t & __lenResult, void * & __result,
+        IDigitalSignatureManager * pDSM,
+        size32_t lenInputdata, const void * inputdata)
+{
+    StringBuffer sbSig;
+    if (pDSM->digiSign(sbSig, lenInputdata, inputdata))
+    {
+        __result = CTXMALLOC(parentCtx, sbSig.length());
+        __lenResult = sbSig.length();
+        memcpy(__result, sbSig.str(), __lenResult);
+    }
+    else
+    {
+        throw MakeStringException(-1, "Unable to create Digital Signature");
+    }
+}
+
+void verifyPKIAlgorithm(const char * pkAlgorithm)
+{
+    if (!strieq(pkAlgorithm, "SHA256"))
+        throw MakeStringException(-1, "Unsupported PKI algorithm (%s) specified", pkAlgorithm);
+}
+
+
+//-----------------------------------------------------------------
+//  Simple cache for instances of Digital Signature Managers
+//-----------------------------------------------------------------
+class CDSMCache
+{
+private:
+    CriticalSection signingDSMCacheCS;//guards modifications to the cache map
+    CriticalSection verifyingDSMCacheCS;//guards modifications to the cache map
+    typedef std::unordered_map <string, IDigitalSignatureManager *> DSMCache;//DSM instances are never deleted, since we are not notified if a module is destroyed
+    DSMCache signingDSMCache;
+    DSMCache verifyingDSMCache;
+
+public:
+    ~CDSMCache()
+    {
+        for ( auto it = signingDSMCache.begin(); it != signingDSMCache.end(); ++it )
+        {
+            ((IInterface *)(*it).second)->Release();
+        }
+        for ( auto it = verifyingDSMCache.begin(); it != verifyingDSMCache.end(); ++it )
+        {
+            ((IInterface *)(*it).second)->Release();
+        }
+    }
+
+    IDigitalSignatureManager * querySigningInstance(const char * algo, const char * privKeyFS, const char * privKeyBuff, const char * passphrase)
+    {
+        VStringBuffer searchKey("%s_%s_%s_%s", algo, privKeyFS, privKeyBuff, passphrase);
+        CriticalBlock block(signingDSMCacheCS);
+        DSMCache::iterator it = signingDSMCache.find(searchKey.str());
+        if (it != signingDSMCache.end())//exists in cache
+        {
+            return (*it).second;
+        }
+        else
+        {
+            IDigitalSignatureManager * pDSM = nullptr;
+            if (!isEmptyString(privKeyFS))
+                pDSM = createDigitalSignatureManagerInstanceFromFiles(nullptr, privKeyFS, passphrase);//uses private key filespec
+            else
+                pDSM = createDigitalSignatureManagerInstanceFromKeys(nullptr, privKeyBuff, passphrase);//uses private key buffer
+
+            signingDSMCache.insert(pair<string, IDigitalSignatureManager*>(searchKey.str(), pDSM));
+            return pDSM;
+        }
+    }
+
+
+    IDigitalSignatureManager * queryVerifyingInstance(const char * algo, const char * pubKeyFS, const char * pubKeyBuff, const char * passphrase)
+    {
+        VStringBuffer searchKey("%s_%s_%s_%s", algo, pubKeyFS, pubKeyBuff, passphrase);
```

Comment:
Do you actually need separate querySigningInstance / queryVerifyingInstance methods? 
And not just a single getDSMInstance method?

They are both passed { algo, [file], [buf], passphrase}, and then code is same.
I'm not sure the code needs to distinguish between the sign/verify when getting an instance, the filename or buffer will be different anyway, and therefore the cache entry/lookup result.